src/plot_api/plot_api.js

```diff
@@ -2508,6 +2527,360 @@ Plotly.relayout = function relayout(gd, astr, val) {
+        if(!list) return;
+        while(list.length) {
+            list.shift();
+        }
+    }
+
+    var aborted = false;
+
+    function executeTransitions() {
+        gd._transitionData._interruptCallbacks.push(function() {
+            aborted = true;
+        });
+
+        // Construct callbacks that are executed on transition end. This ensures the d3 transitions
+        // are *complete* before anything else is done.
+        var numCallbacks = 0;
+        var numCompleted = 0;
+        function makeCallback() {
+            numCallbacks++;
+            return function() {
+                numCompleted++;
+                // When all are complete, perform a redraw:
+                if(!aborted && numCompleted === numCallbacks) {
+                    completeTransition();
+                }
+            };
+        }
+
+        var traceTransitionConfig;
+        var hasTraceTransition = false;
+        var j;
+        var basePlotModules = fullLayout._basePlotModules;
+        var hasAxisTransition = false;
+
+        if(layout) {
+            for(j = 0; j < basePlotModules.length; j++) {
+                if(basePlotModules[j].transitionAxes) {
+                    var newLayout = Lib.expandObjectPaths(layout);
+                    hasAxisTransition = basePlotModules[j].transitionAxes(gd, newLayout, transitionConfig, makeCallback) || hasAxisTransition;
+                }
+            }
+        }
+
+        // Here handle the exception that we refuse to animate scales and axes at the same
+        // time. In other words, if there's an axis transition, then set the data transition
+        // to instantaneous.
+        if(hasAxisTransition) {
+            traceTransitionConfig = Lib.extendFlat({}, transitionConfig);
+            traceTransitionConfig.duration = 0;
+        } else {
+            traceTransitionConfig = transitionConfig;
+        }
+
+        for(j = 0; j < basePlotModules.length; j++) {
+            // Note that we pass a callback to *create* the callback that must be invoked on completion.
+            // This is since not all traces know about transitions, so it greatly simplifies matters if
+            // the trace is responsible for creating a callback, if needed, and then executing it when
+            // the time is right.
+            basePlotModules[j].plot(gd, transitionedTraces, traceTransitionConfig, makeCallback);
+        }
+
+        // If nothing else creates a callback, then this will trigger the completion in the next tick:
+        setTimeout(makeCallback());
+
+        if(!hasAxisTransition && !hasTraceTransition) {
+            return false;
+        }
+
+    }
+
+    function completeTransition() {
+        flushCallbacks(gd._transitionData._interruptCallbacks);
+
+        return Promise.resolve().then(function() {
+            if(transitionConfig.redraw) {
+                return Plotly.redraw(gd);
+            }
+        }).then(function() {
+            gd.emit('plotly_endtransition', []);
```

Comment:
I think it's a really good idea to keep the syntax consistent across these features (even though `plotly_endtransition` makes perfect sense, it might be even easier to remember these features w/ parallel syntax to selection)